### PR TITLE
Emote Panel

### PR DESCRIPTION
**Pull request by larentoun**
## About The Pull Request

This PR adds optional Emote Panel on the sidebar. Supports regular human species, aliens and silicon

Images:
Emotes:
image
image
Silicon/synth version:
image
Game Preferences:
image

P.S. If there are any tips on how to make it prettier (code/visual), be free to comment

## How This Contributes To The Skyrat Roleplay Experience

New people will have an easier time to get accustomed to emotes, and it's much more presentable to read emotes on the sidebar than to use *help

## Changelog

:cl:
add: Emote panel added
/:cl:

### CODE REVIEW

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 1]
```diff
@@ -0,0 +1,978 @@
...
+	var/list/available_emotes = list()
+	var/list/all_emotes = list()
+
+	// code\modules\mob\emote.dm
+	var/list/mob_emotes = list(
+		/mob/proc/emote_flip,
+		/mob/proc/emote_spin
+	)
+	all_emotes += mob_emotes
+	if(istype(src, /mob))
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: use `ismob()`

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 2]
```diff
@@ -0,0 +1,978 @@
...
+		available_emotes += mob_emotes
+
+	// code\modules\mob\living\emote.dm
+	var/list/living_emotes = list(
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: static list

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 3]
```diff
@@ -0,0 +1,978 @@
...
+		/mob/living/proc/emote_whimper,
+		/mob/living/proc/emote_wsmile,
+		/mob/living/proc/emote_yawn,
+		/mob/living/proc/emote_gurgle,
+		/mob/living/proc/emote_inhale,
+		/mob/living/proc/emote_exhale,
+		/mob/living/proc/emote_swear
+	)
+	all_emotes += living_emotes
+	if(istype(src, /mob/living))
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: isliving()

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 4]
```diff
@@ -0,0 +1,978 @@
...
+		available_emotes += living_emotes
+
+	// code\modules\mob\living\carbon\emote.dm
+	var/list/carbon_emotes = list(
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: static

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 5]
```diff
@@ -0,0 +1,978 @@
...
+		/mob/living/carbon/proc/emote_blink_r,
+		/mob/living/carbon/proc/emote_crack,
+		/mob/living/carbon/proc/emote_circle,
+		/mob/living/carbon/proc/emote_moan,
+		///mob/living/carbon/proc/emote_noogie,
+		/mob/living/carbon/proc/emote_slap,
+		/mob/living/carbon/proc/emote_wink
+	)
+	all_emotes += carbon_emotes
+	if(istype(src, /mob/living/carbon))
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: iscarbon

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 6]
```diff
@@ -0,0 +1,978 @@
...
+		available_emotes += carbon_emotes
+
+	// code\modules\mob\living\carbon\human\emote.dm
+	var/list/human_emotes = list(
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: static

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 7]
```diff
@@ -0,0 +1,978 @@
...
+		/mob/living/carbon/human/proc/emote_mumble,
+		/mob/living/carbon/human/proc/emote_pale,
+		/mob/living/carbon/human/proc/emote_raise,
+		/mob/living/carbon/human/proc/emote_salute,
+		/mob/living/carbon/human/proc/emote_shrug,
+		/mob/living/carbon/human/proc/emote_wag,
+		/mob/living/carbon/human/proc/emote_wing
+	)
+	all_emotes += human_emotes
+	if(istype(src, /mob/living/carbon/human))
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: ishuman

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 8]
```diff
@@ -0,0 +1,978 @@
...
+		available_emotes += human_emotes
+
+	// modular_skyrat\modules\emotes\code\emote.dm
+	var/list/skyrat_living_emotes = list(
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: static

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 9]
```diff
@@ -0,0 +1,978 @@
...
+		/mob/living/proc/emote_buzz2,
+		/mob/living/proc/emote_chime,
+		/mob/living/proc/emote_honk,
+		/mob/living/proc/emote_ping,
+		/mob/living/proc/emote_sad,
+		/mob/living/proc/emote_warn,
+		/mob/living/proc/emote_slowclaps
+	)
+	all_emotes += synth_emotes
+	if(istype(src, /mob/living/carbon/human))
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: ishuman

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 10]
```diff
@@ -0,0 +1,978 @@
...
+		available_emotes += brain_emotes
+
+	// code\modules\mob\living\carbon\alien\emote.dm
+	var/list/alien_emotes = list(
+		/mob/living/carbon/alien/proc/emote_gnarl,
+		/mob/living/carbon/alien/proc/emote_hiss,
+		/mob/living/carbon/alien/proc/emote_roar
+	)
+	all_emotes += alien_emotes
+	if(istype(src, /mob/living/carbon/alien))
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: isalien

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 11]
```diff
@@ -0,0 +1,978 @@
...
+		/mob/living/proc/emote_cackle,
+		/mob/living/proc/emote_warble,
+		/mob/living/proc/emote_trills,
+		/mob/living/proc/emote_rpurr,
+		/mob/living/proc/emote_purr,
+		/mob/living/proc/emote_moo,
+		/mob/living/proc/emote_honk1
+	)
+	all_emotes += skyrat_living_emotes
+	if(istype(src, /mob/living))
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: didn't you do a living check before? also, isliving()

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 1 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Appropriate check applied, also moved the checks to make it easier to read

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 2 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Static list applied

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 3 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Appropriate check applied, also moved the checks to make it easier to read

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 4 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Static list applied

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 5 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Appropriate check applied, also moved the checks to make it easier to read

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 6 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Static list applied

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 7 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Appropriate check applied, also moved the checks to make it easier to read

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 8 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Statis list applied

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 9 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Appropriate check applied, also moved the checks to make it easier to read

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 10 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Appropriate check applied, also moved the checks to make it easier to read

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 11 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Changed the check, also moved the checks to make it easier to read

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 12]
```diff
@@ -0,0 +1,987 @@
...
+
+	// code\modules\mob\living\carbon\emote.dm
+	var/static/list/carbon_emotes = list(
+		/mob/living/carbon/proc/emote_airguitar,
+		/mob/living/carbon/proc/emote_blink,
+		/mob/living/carbon/proc/emote_blink_r,
+		/mob/living/carbon/proc/emote_crack,
+		/mob/living/carbon/proc/emote_circle,
+		/mob/living/carbon/proc/emote_moan,
+		///mob/living/carbon/proc/emote_noogie,
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm]: remove commented out code

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 13]
```diff
@@ -0,0 +1,987 @@
...
+
+	// Clearing all emotes before applying new ones
+	verbs -= all_emotes
+
+	// Checking if preferences allow emote panel
+	if(!src.client?.prefs?.read_preference(/datum/preference/toggle/emote_panel))
+		return
+
+	// Checking emote availability
+	if(isbrain(src))
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm:218]: this looks _much_ better

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm [hunk 14]
```diff
@@ -0,0 +1,987 @@
...
+	set category = "Emotes"
+	usr.emote("sniff", intentional = TRUE)
+
+/mob/living/proc/emote_stare()
+	set name = "~ Stare"
+	set category = "Emotes"
+	usr.emote("stare", intentional = TRUE)
+
+/mob/living/proc/emote_strech()
+	set name = "~ Stretch"
```
Comment (Zonespace27) [inline modular_skyrat/modules/emote_panel/code/emote_panel.dm:462]: what's the difference between the >, ~, and | | emotes?

Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm:462]: That's basically only used for ordering. 
\> Emote has sound effect
| Emote | is an "action" (flip, kiss)
< Emote > is race specific (basically only used on synth, silicon and brain. Could be used on simple mobs in the future)
~ Emote is a regular text-only emote

**Comment by desertmote** [comment]
Maybe add something for the overlay emotes?
https://github.com/Skyrat-SS13/Skyrat-tg/tree/master/modular_skyrat/modules/emotes/code/additionalemotes

**Comment by larentoun** [comment]
> Maybe add something for the overlay emotes? https://github.com/Skyrat-SS13/Skyrat-tg/tree/master/modular_skyrat/modules/emotes/code/additionalemotes

Hmm, didn't quite catch what were these emotes about. After reading for the second time, I will think of something

**Comment by Iamgoofball** [comment]
we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely

there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel

File: modular_skyrat/modules/emote_panel/code/emote_panel.dm (hunk 12 above)
Comment (larentoun) [reply modular_skyrat/modules/emote_panel/code/emote_panel.dm]: Removed the proc and the commented code

**Comment by larentoun** [comment]
> we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> 
> there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel

I see. Still, parts of this code could be reused for an emote wheel. And, still, there is only one presentable option to look up the emotes, and it's in the keybindings. *help doesn't really help for new players, so that verb panel may be helpful for them. After they will be accustomed, they may use the chat or keybindings for it.

I will probably look how to make a proper radial menu for it after this PR... Since that menu can be and will be cluttered...

**Comment by hugmeorelse** [comment]
> we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> 
> there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel

I have to agree with Goofball's assessment. This is a lot of clutter, and it doesn't really offer an advantage to us personally unfortunately.

> > we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> > there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel
> 
> I see. Still, parts of this code could be reused for an emote wheel. And, still, there is only one presentable option to look up the emotes, and it's in the keybindings. *help doesn't really help for new players, so that verb panel may be helpful for them. After they will be accustomed, they may use the chat or keybindings for it.
> 
> I will probably look how to make a proper radial menu for it after this PR... Since that menu can be and will be cluttered...

I understand you're bringing this PR upstream to minimize any possible disruption in the form of merge conflicts on your end which is completely understandable. We've been there, however, I feel like there's a more elegant solution here we can go for. An emote wheel would be interesting. Someone, at one point, worked on something like that. We've had that formerly for the Bay-style Aiming PR that one of our former maintainers worked on. You could possibly recycle the code from there unless it was added in later by our aforementioned maintainer: #800 It's been a few years.

As is, I don't really see this as something we need upstream, since Skyrat is all about typing it out rather than using macros as a shortcut for emotes. However, I do understand you're looking to cater to a more general audience. Is this something that is commonly used in your community's other servers?

**Comment by hugmeorelse** [comment]
> > we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> > there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel
> 
> I have to agree with Goofball's assessment. This is a lot of clutter, and it doesn't really offer an advantage to us personally unfortunately.
> 
> > > we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> > > there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel
> > 
> > 
> > I see. Still, parts of this code could be reused for an emote wheel. And, still, there is only one presentable option to look up the emotes, and it's in the keybindings. *help doesn't really help for new players, so that verb panel may be helpful for them. After they will be accustomed, they may use the chat or keybindings for it.
> > I will probably look how to make a proper radial menu for it after this PR... Since that menu can be and will be cluttered...
> 
> I understand you're bringing this PR upstream to minimize any possible disruption in the form of merge conflicts on your end which is completely understandable. We've been there, however, I feel like there's a more elegant solution here we can go for. An emote wheel would be interesting. Someone, at one point, worked on something like that. We've had that formerly for the Bay-style Aiming PR that one of our former maintainers worked on. You could possibly recycle the code from there unless it was added in later by our aforementioned maintainer: #800 It's been a few years.
> 
> As is, I don't really see this as something we need upstream, since Skyrat is all about typing it out rather than using macros as a shortcut for emotes. However, I do understand you're looking to cater to a more general audience. Is this something that is commonly used in your community's other servers?

**Comment by larentoun** [comment]
> > we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> > there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel
> 
> I have to agree with Goofball's assessment. This is a lot of clutter, and it doesn't really offer an advantage to us personally.
> 
> > > we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> > > there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel
> > 
> > 
> > I see. Still, parts of this code could be reused for an emote wheel. And, still, there is only one presentable option to look up the emotes, and it's in the keybindings. *help doesn't really help for new players, so that verb panel may be helpful for them. After they will be accustomed, they may use the chat or keybindings for it.
> > I will probably look how to make a proper radial menu for it after this PR... Since that menu can be and will be cluttered...
> 
> I understand you're bringing this PR upstream to minimize any possible disruption in the form of merge conflicts on your end which is completely understandable. We've been there, however, I feel like there's a more elegant solution here we can go for. An emote wheel would be interesting. Someone, at one point, worked on something like that. We've had that formerly for the Bay-style Aiming PR that one of our former maintainers worked on. You could possibly recycle the code from there: #800
> 
> As is, I don't really see this as something we need upstream, since Skyrat is all about typing it out rather than using macros as a shortcut for emotes. However, I do understand you're looking to cater to a more general audience. Is this something that is commonly used in your community's other servers?

Yeah, I'm from SS220 community and for a long time we had Emote Panel, and most people there, who attended on our closed Skyrat testing and played on TG, were missing it.

It's no problem for us to do it modularly (since we will be doing some our own stuff modularly anyway), but, perhaps, it would be a good addition to starting out players.

Thanks for replying! If that won't go upstream, in the future, I hope, I will come here with a more elegant solution.

**Comment by hugmeorelse** [comment]
> > > we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> > > there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel
> > 
> > 
> > I have to agree with Goofball's assessment. This is a lot of clutter, and it doesn't really offer an advantage to us personally.
> > > > we should not be adding more stuff to an already cluttered verbs panel, the end goal is to remove it entirely
> > > > there should be a button to bring up an emote wheel and you should be able to select that way with prefs to configure the wheel
> > > 
> > > 
> > > I see. Still, parts of this code could be reused for an emote wheel. And, still, there is only one presentable option to look up the emotes, and it's in the keybindings. *help doesn't really help for new players, so that verb panel may be helpful for them. After they will be accustomed, they may use the chat or keybindings for it.
> > > I will probably look how to make a proper radial menu for it after this PR... Since that menu can be and will be cluttered...
> > 
> > 
> > I understand you're bringing this PR upstream to minimize any possible disruption in the form of merge conflicts on your end which is completely understandable. We've been there, however, I feel like there's a more elegant solution here we can go for. An emote wheel would be interesting. Someone, at one point, worked on something like that. We've had that formerly for the Bay-style Aiming PR that one of our former maintainers worked on. You could possibly recycle the code from there: #800
> > As is, I don't really see this as something we need upstream, since Skyrat is all about typing it out rather than using macros as a shortcut for emotes. However, I do understand you're looking to cater to a more general audience. Is this something that is commonly used in your community's other servers?
> 
> Yeah, I'm from SS220 community and for a long time we had Emote Panel, and most people there, who attended on our closed Skyrat testing and played on TG, were missing it.
> 
> It's no problem for us to do it modularly (since we will be doing some our own stuff modularly anyway), but, perhaps, it would be a good addition to starting out players.
> 
> Thanks for replying! If that won't go upstream, in the future, I hope, I will come here with a more elegant solution.

I'll talk to our maintainers and see what we can do. We could possibly merge this temporarily out of good will with the expectation that this will not be a permanent temporary solution, but I'll have to discuss the matter with them since they're the ones working to the maintain the codebase. I'm ultimately leaving the decision with them.

Is this made as a preference by the way? If so, that may possibly help. Also, as a heads up, if you do make a more elegant solution, I recommend bringing it upstream to TG (I don't think they'll accept this PR as is upstream, so it would have to be something like the emote wheel and so forth with less clutter). They tend to have solid standards, and it may be to their interest to have something like this themselves. Making the change there should minimize disruption on Skyrat too, thereby minimizing disruption on your server as well. They may also have some suggestions.

**Comment by larentoun** [comment]
> 
> I'll talk to our maintainers and see what we can do. We could possibly merge this temporarily out of good will with the expectation that this will not be a permanent temporary solution, but I'll have to discuss the matter with them since they're the ones working to the maintain the codebase. I'm ultimately leaving the decision with them.
> 
> Is this made as a preference by the way? If so, that may possibly help. Also, as a heads up, if you do make a more elegant solution, I recommend bringing it upstream to TG (I don't think they'll accept this PR as is upstream, so it would have to be something like the emote wheel and so forth with less clutter). They tend to have solid standards, and it may be to their interest to have something like this themselves. Making the change there should minimize disruption on Skyrat too, thereby minimizing disruption on your server as well. They may also have some suggestions.

Yeah, it's an optional feature with a preference, which is defaulted to false.

I still don't have a clear vision how to make it elegant with an emote wheel, since it would be smart to attach preferenced emotes to a wheel - not all of them. But here my overall experience (I'm barely better than a beginner) fails me, so would need to read more about radial menus.

The mission goal is to show what emotes can do what, since *help doesn't help to differentiate between usual emotes, overlay emotes and audible emotes.

If maintainers will see any performance issues (I hope there is close to none)/clutterness be a problem, then I'm fine with them not merging to upstream. I've already tried it to modularly add to our fork, and, it seems, no merging problems appeared (unless dme file will change, not sure how merging problems appear, due to, again, my lack of experience).

If anything, I can give contacts with our main host/contributor in discord

**Review by Gandalf2k15 — APPROVED** [review]
Code is fine, assuming we can't find a better way, but given this is a preference, and one I would use, it's good.

**Pull request merged**